### ISSUE TITLE
Support for GL_OES_EGL_image_external on Android?

### ISSUE THREAD

**Issue by oviano** [floooh/sokol#406]
As in, when using mediacodec video decoding, you end up with one of these textures, which needs samplerExternalOES.

What would be involved to support this, just additions to sokol_gfx.h or also to the shader tool? All I want to do is inject one of these textures into an sg_image, as I am doing similarly with Metal and D3D11 textures and make it work with the right shader/sampler.

I’ve been porting my video streaming project to Sokol, having been using bgfx until now. Using a combination of sokol_gfx and sokol_imgui instead of bgfx and nanoVG, I’ve pretty much got things working.

I have hardware accelerated video decoding working on Apple platforms and Windows (d3d11 requires a small mod to sokol_gfx for which I will make a PR soon), but I think I may have hit a brick wall with Android as I’m unsure what I need to do to be able to use this type of texture.

This gist gives a few hints.

https://gist.github.com/rexguo/6696123

And bgfx had support at one point via this PR (later reverted in the main branch), which works for me:

https://github.com/bkaradzic/bgfx/commit/f07a0f51792db2351f6a38e1625a704be159a6ee

By the way, I’m really loving using sokol, it’s so nice and compact and free of bloat.

**Comment by floooh** [comment]
Hmm... if it would just involve some additional GL extensions, then this wouldn't be a problem, but I really don't want to bring any EGL or even Android-specific code into core sokol-gfx (like this Android GraphicsBuffer thing).

If it would just be about creation a special texture type and updating the texture data with a special function, then one could use "injected" native textures into sokol-gfx (like in this example which uses GL buffers and textures created outside of sokol-gfx: https://github.com/floooh/sokol-samples/blob/master/glfw/inject-glfw.c).

But the problem is the special texture target type GL_TEXTURE_EXTERNAL_OES, this seeps into a lot of places in the GL backend :( 

One could create an 'extension header' with additional sokol-gfx functions and which implement the additionally required features, for instance:

```
sg_bla_make_image(...)
sg_bla_destroy_image(...)
sg_bla_image_ext(...)
sg_bla_apply_bindings(...)
```
...not sure if this is enough, but you get the idea...such an extension header would be included after sokol_gfx.h, and that way the implementation can directly access all internal data and functions of sokol_gfx.h.

**Comment by floooh** [comment]
...on further though, a "non-disruptive" option building on the "resource injection idea" which might allow such external texture types, but without bringing any EGL or Android-specific code into sokol-gfx might be not just injecting an external GL texture handler, but also the texture target type (GL_TEXTURE_EXTERNAL_OES vs GL_TEXTURE_2D).

Internal GL backend textures already have the target type stored here:

https://github.com/floooh/sokol/blob/345fa62a3d8f43b11b404714fe84b203d4536feb/sokol_gfx.h#L2955-L2961

...this is set in sg_make_image(), and then used in all GL functions which need the target type.

This target type could be "injected" just as the GL texture handle in ```sg_image_desc```:

https://github.com/floooh/sokol/blob/345fa62a3d8f43b11b404714fe84b203d4536feb/sokol_gfx.h#L1630

...so in addition to ```gl_textures[]``` there would be a ```gl_texture_target``` entry which would need to be set to GL_TEXTURE_EXTERNAL_OES. If this idea works, then such special textures would need to be created and updated outside of sokol-gfx, but sokol-gfx still could use them as injected texture resources for rendering... this sounds as if it could work...

**Comment by oviano** [comment]
In my scenario the texture is created entirely outside sokol, by the mediacodec video decoding engine. The way I had it working with bgfx was that I just used the override_internal (sokol "injected" equivalent) to pass in this texture.

Then bgfx was able to recognise it was GL_TEXTURE_EXTERNAL_OES (not sure how) and adjust accordingly.

I was having a quick look earlier at the sokol code and I suppose in this case I'd want _sg_gl_texture_target to return GL_TEXTURE_EXTERNAL_OES, so maybe you could have a generic SG_IMAGETYPE_EXTERNAL.

The other thing I noticed is that sokol-shadc won't compile the following shader, which I think *should* work, as your utility seems to be based on recent enough glsl code to support the extension:

```/* ext vertex shader */
@vs ext_vs
in vec4 position;
in vec2 texcoord0;
out vec2 uv;

void main()
{
    gl_Position = position;
    uv = texcoord0;
}
@end

/* ext fragment shader */
@fs ext_fs
#extension GL_OES_EGL_image_external : require
uniform samplerExternalOES ext_fs_tex;

in vec2 uv;
out vec4 frag_color;

void main()
{
	frag_color = texture(ext_fs_tex, uv);
}
@end

/* ext shader program */
@program ext ext_vs ext_fs
```

When I compile this for glsl100 it first of all complains that the extension required is GL_OES_EGL_image_external_essl3 not GL_OES_EGL_image_external. I think this might be something out of your control though, and I don't know if that is an issue.

Secondly, if you replace it with GL_OES_EGL_image_external_essl3 it then complains it cannot find a match for the texture() function with these parameters. I also tried texture2D, and checked the glsl source and it should work.

I can maybe just bypass sokol-shadc for this shader, and hopefully this should work:

```
    #version 100
    #extension GL_OES_EGL_image_external : require
    precision mediump float;
    precision highp int;
    
    uniform highp samplerExternalOES ext_fs_tex;
    
    varying highp vec2 uv;
    
    void main()
    {
        gl_FragData[0] = texture2D(ext_fs_tex, uv);
    }
````

**Comment by floooh** [comment]
PS: on further-further though this additional ```gl_texture_target``` item is also needed anyway for using any injected textures that are not of the GL_TEXTURE_2D type... so this addition makes sense anyway, in fact it's an oversight that this doesn't exist...

**Comment by oviano** [comment]
Yes, that all makes perfect sense to me.

**Comment by floooh** [comment]
...ok sounds good... wanna prepare a PR with this sg_image_desc.gl_texture_target item? It should be pretty straightforward. If you do, please keep the other D3D related changed in a separate PR though :)

**Comment by oviano** [comment]
Yes, I'll have a go at this, and prepare separate PRs for my changes.

In terms of the D3D11 one this is a case of making the injected d3d11_texture a more generic d3d11_resource, and then at the point it is used, it makes a call to QueryInterface to determine if it is, in fact, a shader resource view. If it is, then it can set the srv from this, rather than create a new one. In my case it's a video decoded texture in NV12 format, so I create an SRV for each of the two planes, and pass these to sokol instead of the texture.

It's the exact same approach I used for bgfx, and it works fine in sokol too, and allows me to use shaders with injected D3D11 textures.

**Comment by oviano** [comment]
My pull requests:

https://github.com/floooh/sokol/pull/409
https://github.com/floooh/sokol/pull/410
https://github.com/floooh/sokol/pull/409

**Comment by floooh** [comment]
Closing this ticket because #411 has been merged (next is #410).
